L58-python-async-api.md

```diff
@@ -0,0 +1,432 @@
+After all, it is not a good practice to introduce uncertainty (return coroutine
+or regular object) to our API. This decision is made to respect our contract of
+API since GA in 2016, including the underlying behaviors. **Developers who use
+our current API should not be affected by this effort.**
+
+For users who want to migrate to new API, the granularity migration is per
+channel, per server level. For both channel-side and server-side, the adoption
+of `asyncio` does not only include cosmetic changes of adding `async`/`await`
+keywords throughout the code base, it also requires thinking about the change
+from a potentially multi-threaded application to a single-threaded asynchronous
+application. For example, the synchronization methods have to be changed to use
+the `asyncio` edition (see [Synchronization
+Primitives](https://docs.python.org/3/library/asyncio-sync.html)).
+
+### Overview of Async API
+
+All `asyncio` related API will be kept inside the `grpc.aio` module. Take the
+most common use case as an example, the new API will diverge from the old API
+since the creation of `Channel` and `Server` object. New creation methods are
+added to instantiate async `Server` or `Channel` object:
+
+```Python
+import grpc
+server = grpc.aio.server(...)
+channel = grpc.aio.insecure_channel(...)
+channel = grpc.aio.secure_channel(...)
+channel = grpc.aio.intercept_channel(...)
+```
+
+To reduce cognitive burden, the new `asyncio` API should share the same
+parameters as the current API, and keep the usage similar, except the following
+cases:
+
+* `grpc.aio.server` no longer require the application to provide a
+  `ThreadPoolExecutor`;
+* Interfaces returning `Future` object are replaced.
+
+### Demo Snippet of Async API
+
+Server side:
+```Python
+class AsyncGreeter(helloworld_pb2_grpc.GreeterServicer):
+
+    async def SayHello(self, request, context):
+        await asyncio.sleep(1)
+        return helloworld_pb2.HelloReply(message='Hello, %s!' % request.name)
+
+server = grpc.aio.server()
+server.add_insecure_port(":50051")
+helloworld_pb2_grpc.add_GreeterServicer_to_server(Greeter(), server)
+server.start()
+await server.wait_for_termination()
+```
+
+Client side:
+```Python
+async with grpc.aio.insecure_channel("localhost:50051") as channel:
+    stub = echo_pb2_grpc.EchoStub(channel)
+    response = await stub.Hi(echo_pb2.EchoRequest(message="ping"))
+```
+
+### Channel-Side
+
+Changes in `grpc.aio.Channel`:
+* The object returned by `unary_unary`, `unary_stream`, `stream_unary`, and
+  `stream_tream` will be `MultiCallable` objects in `grpc.aio`.
+
+Changes in `grpc.aio.Call`:
+* All methods return coroutine object.
+
+Changes for `MultiCallable` classes in `grpc.aio`:
+* `grpc.aio.UnaryUnaryMultiCallable` and `grpc.aio.StreamUnaryMultiCallable`
+  returns `Awaitable` object for `__call__`;
+* `grpc.aio.UnaryUnaryMultiCallable` and `grpc.aio.StreamUnaryMultiCallable`
+  returns `(Awaitable, grpc.aio.Call)` for `with_call`;
```

Comment:
For the sake of simplicity and for having an orthogonal API with the synchronous one I would go for having a `with_call` with an interface like this one

```python
response, call = await stub.Hi.with_call(...)
task = asyncio.ensure_future(stub.Hi.with_call(...))
..
..
response, call = task.result()
```

So returning basically an _awaitable_ object implemented by the coroutine, and having the coroutine the ability to return a tuple with two objects when it has finished.

IMO, if I'm not missing something, this would implement the same contract as the synchronous one and will make the implementation under the hood easier since its only a matter of returning two parameters instead of one when the `with_call` is used.

Regarding the object to return for the call, we might return the same object `grpc.aio.Call`  that we could be using for the `future()` method but without yielding if the `await` expr is used, take a look to the `Future` implementation of Asyncio [1].

The following should work, I did not test it, and the user should not suffer any context switch:

```python
response, call = await stub.Hi.with_call(...)
await call
print(call.status)
```

Also a bit related, we would need to implement the cancelation of a task, which would impact the `with_call` and the `__call__` methods.

For example this should be doable:

```python
task = asyncio.ensure_future(stub.Hi.with_call(...))
task.cancel()
```

Which theoretically should be implementable by changing a bit the `call.pyx.pxi` [2] file with something like

```python
   try:
       await self._waiter_call
   except CancelledError:
       if not self._canceled:
           # Behind the scenes ends up by calling the grpc_call_cancel
           # gRPC method
           self.cancel()
    .....
```

Later on, this cancellation mechanism might be used by the `grpc.aio.Call.cancel` the method, which behind the scenes would cancel the task.

[1] https://github.com/python/cpython/blob/master/Lib/asyncio/futures.py#L260

[2] https://github.com/Skyscanner/grpc/pull/1/files#diff-d07d1d5928e4bc4defeb2e1218c57186R136

Reply:
I hope we can merge the `__call__`, `with_call` and `futures`. It can be much simpler. By default, the invocation to the stub will return an `asyncio.Task` object. Users can control the life cycle of that RPC by invoking methods on that object. On the other hand, if users just care about the result, they can `await` on the `Task`.

```Python
# Normal call
response = await stub.Hi(HiRequest)

# In the case of cancel
call = stub.Hi(HiRequest)
if number_of_greeting > 0:
    call.cancel()
else:
    response = await call
```

Reply:
Merging `__call__`, `with_call` would mean not having the same implementation btw the sync and async version of the API, are we ok with that?

If so, I kinda agree with your proposal, but adding the following modifications:

1) The three original methods for making a call would be implemented with just one way. So `__call__`, `with_call` and `future` implemented under the `__call__` method
leveraging on the _awaitable_ property of a `grpc.aio.Call` object, so the following expressions would be valid:

```python
response = await stub.Hi(...)
stub.Hi(...).cancel()
asyncio.wait_for(stub.Hi(...), timeout=10)
```

2) I would move everything that is not related to the call managing from the `grpc.aio.Call` object under a new `grpc.aio.Response` object. In the synchronous version,
the `_Rendezvous` [1] implements both features, managing the call and returning properties of the response. If we do not change this, with the proposal of unifying the `with_call` method with the `__call__` the user will be forced to repeat the following pattern:

```python
call = stub.Hi(...)
response = await call
print(response)
print(call.status)
```

The alternative with a dedicated response object would be:

```python
response = await stub.Hi(...)
print(response.content)
print(response.code)
```


[1] https://github.com/Skyscanner/grpc/blob/7b0576787898ee670c7a56b3fac34acfb1a62df0/src/python/grpcio/grpc/_channel.py#L251

Reply:
Glad you agree with modification 1.  Your modification 2 is a good catch that I overlooked the situation when users want to check `status-code` and `status-message`. It is a valid option; many frameworks are using it.

A slight drawback is that if we are going to add a `grpc.aio.Response`, we might want to also add `grpc.aio.Request`, `grpc.aio.ResponseIterator` etc.. To start a new pattern, we should be cautious and discuss them thoroughly.

To solve this problem, this is my proposal. In current design, if a call finishes abnormally (status not OK), an exception will be raised. We can use the same pattern in the new API.

```Python
try:
    response  = await stub.Hi(...)
catch grpc.RpcError as rpc_error:
    print(rpc_error.code())
    print(rpc_error.details())
```

Reply:
Is there any other attribute that the user might need when there is no error? What about initial and trailing metadata [1]?

So just for summarizing the output of the discussion:

1) Won't be any more `__call__`, `with_call` and `future` methods, everything will be handled using the unique `__call__` available.
2) The `__call__` method will return a `grpc.aio.Call` object which would be _awaitable_, returning the body once the task is done.

am I right? If the metadata and trailing metadata are not an issue, I'm completely Ok with that proposal. 


[1] https://github.com/Skyscanner/grpc/blob/7b0576787898ee670c7a56b3fac34acfb1a62df0/src/python/grpcio/grpc/_channel.py#L387

Reply:
Metadata does matter.

As you suggested, the metadata can be accessed with slightly different way. It separate the receiving of metadata and the entire response message. E.g. `with_call` returns only after the RPC is finished. It prevents users from reacting to `initial_metadata` if received early.

```Python
call = stub.Hi(...)
if validate(call.initial_metadata()):
    response = await call
else:
    call.cancel()
```

IMO this proposal should facilitate the usage to `Call` objects since the logic suits `asyncio` environment better.

Reply:
umm, you mean:

```python
initial_metadata = await call.initial_metadata()
```

?

Reply:
Good catch!